src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2423,9 +2423,11 @@ void ShenandoahHeap::set_concurrent_young_mark_in_progress(bool in_progress) {
 void ShenandoahHeap::set_concurrent_old_mark_in_progress(bool in_progress) {
 #ifdef ASSERT
   // has_forwarded_objects() iff UPDATEREFS or EVACUATION
-  bool has_forwarded = has_forwarded_objects()? 1: 0;
-  bool updating_or_evacuating = _gc_state.is_set(UPDATEREFS | EVACUATION)? 1: 0;
-  assert (has_forwarded == updating_or_evacuating, "Has forwarded objects iff updating or evacuating");
+  bool has_forwarded = has_forwarded_objects();
+  bool updating_or_evacuating = _gc_state.is_set(UPDATEREFS | EVACUATION);
+  bool evacuating = _gc_state.is_set(EVACUATION);
+  assert ((has_forwarded == updating_or_evacuating) || (evacuating && !has_forwarded && collection_set()->is_empty()),
```

Comment:
Sorry, I did not realize this is generational Shenandoah. I am a bit puzzled why setting `EVACUATION` flag if collection set is empty?

Reply:
It's a trick we use for a form of "abbreviated" cycle.  Sometimes, we have nothing to evacuate, but we have a bunch of regions that reside in young, which are old enough to be promoted.  If they are populated with a sufficient amount of live data, we promote them "in place" by simply relabeling the affiliation of the region rather than evacuating it.  But there's a non-trivial amount of processing required on the content of the region when it becomes old, so that the region can now maintain a remembered set of pointers from old to young.  We normally do this processing as part of the evacuation cycle.  In very rare scenarios, we'll end up with an evacuation cycle that only deals with promote-in-place regions, and nothing else...  That's the special scenario we're talking about here.

Reply:
Thanks for the explanation!